### DISCUSSION TITLE
How can I attach authenticated users to records, like the creator of a post?

### DISCUSSION THREAD

**Discussion by Allyedge** [pocketbase/pocketbase#200] [Q&A] [answered]
Hey! I am currently trying out Pocketbase and I am trying to make my example posts be attached to a user. But if I specify that user with an ID, I feel like it doesn't really fit in with the whole authorization logic.

Is there a way to create a record and attach a user to them? I am using the REST API and not the SDK.

The current way it works.

```
POST -> {
    "title": "test",
    "user": "some_id"
}
```

How I would like it to work.

```
POST -> {
    "title": "test",
    "user": "user_by_the_authorization_token_from_the_header"
}
```

Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
Assuming that the `user` from your above example is  a `User relation` field, then Yes, you'll have to pass the current user id to the post data in order to "link/attach" the record to the user.

Why do you "feel like it doesn't really fit in with the whole authorization logic"? Depending on your collection rules, you can even allow guests to perform update/create.

If we do the attachment in the background by default, this may not be expected by everyone because you can have multiple user fields, you may want only user with specific profile field (eg. role) to be able to set/change the "author" of the post, etc.). It is just too ambiguous to have it integrated by default and it is better to be explicit in this type of situations.

Of course, if you use PocketBase as framework, then you could do the attachment automatically with for example a [OnRecordBeforeCreateRequest hook](https://pocketbase.io/docs/event-hooks/#record-hooks).

**Reply by Allyedge** [reply]
Well, yeah but my use case is rather simple. At least for now.

Just like how when we talk here, our replies are attached to our users, is there an easier way to do it in Pocketbase?

Perhaps I should use a relation instead of the user type.

**Reply by ganigeorgiev** [reply]
The `relation` type field also requires to pass a record id.
I understand that for simple use cases this may be bothersome and boilerplate-ish, but as I mentioned there isn't an easy non-ambiguous way (without hooks) to make the assignment automated.

For me, passing the user id you want to associate is intuitive, easy and straightforward enough.

I'm eager to change it, but we all have first to agree on:
- what happens if a guest is allowed to create/update?
- what happens if an admin create/update the record?
- what happens if the `user` type field is optional - do we always assign the authenticated user?
- what happens if the `user` type field allows multiple users (aka. having more than one author)?
- what happens if the record has more than one `user` type field (eg. author, editor, maintainer, etc.)?
- ...

If you need custom default behavior, my recommendation is to use the available hooks (currently the documentation is a bit lacking on this, but more advanced usage guides and examples are already in the roadmap).

**Reply by Allyedge** [reply]
Is passing a user ID safe, though? Can't anyone with access to the API then create records with a random user attached?

**Reply by ganigeorgiev** [reply]
It is up to the developer to manage the collection permissions.

For example, if you want to allow the user to submit a post with only their id, you could use the following create/update action rule:
```
// assuming that `user` is a single required `user` type field
user = @request.user.id
```

**Reply by Allyedge** [reply]
Oh, that might be what I'm looking for, thanks!

Awesome project, by the way. I will definitely keep experimenting. 🙂

**Reply by ganigeorgiev** [reply]
@Allyedge One small addition - the above rule on create will prevent the user to submit a different user id, because the rule is applied after the record insertion (we are doing it in a transaction, and if the rule fails we just return an error that it cannot be created).

But on update the user will be allowed to assign it to someone else, because on update the rule execute before applying the changes, aka. will act as access rule only. And therefore, to additionally validate the submitted data, you could make use of the `@request.data` filter:
```
user = @request.user.id && (@request.data.user = "" || @request.data.user = @request.user.id)
```
- `user = @request.user.id` ensures that only the post owner/author is allowed to update the record
- `@request.data.user = ""` allows submitting update requests without specifying user id, since the update action is PATCH and you can send only the data you want to update
- `@request.data.user = @request.user.id` - ensures that, in case a user id is sent in the update request, it should always match with the current authenticated user.

**Reply by Allyedge** [reply]
Good to know, thanks again! 👍

**Reply by IceWreck** [reply]
Thanks, I was looking for rules like these. You should add these example rules in the documentation, especially the PATCH one.

Also, I want to thank you for building pocketbase and I can see you've put in immense thought into making it work as a go library and I really appreciate that. I'm trying it out for a university project and it lets me leave the CRUD parts to pocketbase so I can focus on the business logic myself.

**Reply by rostrovsky** [reply]
Agree with the above statement. Documenting a list of most common examples / recipes that are used almost 100% of the time would be extremely helpful e.g.

#### Prevent guest access
```
@request.user.id!=null || @request.user.id!=""
```

#### Allow viewing only user's items
```
user = @request.user.id
```

#### Above combined
```
user = @request.user.id && (@request.user.id!=null || @request.user.id!="")
```

**Reply by ganigeorgiev** [reply]
@rostrovsky Yes, the docs will be updated in the near future with more examples and guides.

_btw. you don't need to check for both `null` and empty string because they will be COALESCE to the same value, aka. they do the same thing (this is true since [PocketBase v0.3.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.3.0) because we no longer store `null` values and any `null` value is casted to the field zero-default, which for the id is empty string)._
